src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_process_net_command_json.py

```diff
@@ -460,15 +461,41 @@ def on_source_request(self, py_db, request):
         :param SourceRequest request:
         '''
         source_reference = request.arguments.sourceReference
-        content = ''
+        server_filename = None
+        content = None
+
         if source_reference != 0:
             server_filename = pydevd_file_utils.get_server_filename_from_source_reference(source_reference)
-            if os.path.exists(server_filename):
-                with open(server_filename, 'r') as stream:
-                    content = stream.read()
+            if server_filename:
+                # Try direct file access first - it's much faster when available.
+                try:
+                    with open(server_filename, 'r') as stream:
+                        content = stream.read()
+                except:
+                    pass
+
+                if content is None:
+                    # File might not exist at all, or we might not have a permission to read it,
+                    # but it might also be inside a zipfile, or an IPython cell. In this case,
+                    # linecache might still be able to retrieve the source.
+                    lines = (linecache.getline(server_filename, i) for i in itertools.count(1))
```

Comment:
I think you could use `linecache.getlines` to get the whole file at once instead of getting it line-by-line.

Reply:
It's not a part of its public API - at least I don't see it there:
https://docs.python.org/3/library/linecache.html

Reply:
Humm, I see it's not in the docs, but it's been there since the first revision up to the latest master (so, I don't think they'd break it).

Anyways not sure... I think I'd go with the `linecache.getlines` -- even not being in the docs as it should be faster and have less issues with threads (as theoretically the cache could be invalidated in a thread while we're reading it) , but I'll leave that up to you whether you think that's better or not.

Reply:
In practice in most cases we'd just go straight for the disk file anyway, so perf is unlikely to be an issue. For IPython, it hacks the cache so that it's not really a cache anymore (i.e. the entries for cells never expire, and are effectively immutable so long as the cell is alive).